### PR TITLE
refactor graph node names

### PR DESCRIPTION
**Pull request by deads2k**
It's a little hard to follow exactly how the different nodes are named and failure to standardize has produced some strange collisions in the general case.  This refactors the unique name determination to be easier to discover.

@smarterclayton I needed this to start adding secrets and service accounts in for bad pod templates.

### CODE REVIEW

File: pkg/api/graph/types.go [hunk 1]
```diff
@@ -47,6 +51,73 @@ const (
 	ReferencedImageLayerGraphEdgeKind
 )
 
+var kindToGraphKind = map[reflect.Type]interface{}{}
+
+func init() {
+	kindToGraphKind[reflect.TypeOf(&image.ImageStream{})] = ImageStreamGraphKind
+	kindToGraphKind[reflect.TypeOf(&image.ImageStream{})] = ImageStreamGraphKind
```
Comment (ncdc) [inline pkg/api/graph/types.go]: duplicate line?

Comment (liggitt) [inline pkg/api/graph/types.go]: ha, I was staring at that, trying to figure out what reflection magic required this to be set twice...

Comment (ncdc) [inline pkg/api/graph/types.go]: because image streams are so cool!

Comment (deads2k) [inline pkg/api/graph/types.go]: > duplicate line?

removed.

**Comment by deads2k** [comment]
@smarterclayton I'm going to have a series of refactors as I try to split the graph package apart to allow for easier composition of types.  Do you have another victim you'd like to review?

**Comment by smarterclayton** [comment]
Jessica did the first review but she probably doesn't want to continue it.  Andy is probably best.

**Comment by ncdc** [comment]
Yeah, I can do it

**Comment by deads2k** [comment]
> Yeah, I can do it

Sucker :)

So here's the vision.  I want to split the single graph package apart by functional area along the `api/types` boundaries.  Each `functionarea/graph` package will contain types to make the nodes and functions to create one generation of edges on a particular type.  Splitting it apart should make it easier to inspect what we have an unit test the smaller pieces.

I needed this pull just to start getting my head around what is using and pulling references to what.

File: pkg/api/graph/types.go [hunk 2]
```diff
@@ -47,6 +51,72 @@ const (
...
+func GetUniqueNamespaceNodeName(obj runtime.Object) UniqueName {
+	return getUniqueNodeName(obj, true)
+}
+
+func GetUniqueRootScopedNodeName(obj runtime.Object) UniqueName {
+	return getUniqueNodeName(obj, false)
+}
+
+func getUniqueNodeName(obj runtime.Object, namespaced bool) UniqueName {
+	objType := reflect.TypeOf(obj)
```
Comment (smarterclayton) [inline pkg/api/graph/types.go]: How is this better than a switch?

Comment (deads2k) [inline pkg/api/graph/types.go]: > How is this better than a switch?

A map allows for future inspection of registered keys so that we can figure out how many api resources cannot be referenced in the map.

Comment (smarterclayton) [inline pkg/api/graph/types.go]: Honestly, that's a pretty weak argument.  Test cases can do that too.

File: pkg/api/graph/types.go [hunk 3]
```diff
@@ -47,6 +51,72 @@ const (
...
+	kindToGraphKind[reflect.TypeOf(&image.Image{})] = ImageGraphKind
+	kindToGraphKind[reflect.TypeOf(&build.BuildConfig{})] = BuildConfigGraphKind
+	kindToGraphKind[reflect.TypeOf(&build.Build{})] = BuildGraphKind
+	kindToGraphKind[reflect.TypeOf(&deploy.DeploymentConfig{})] = DeploymentConfigGraphKind
+	kindToGraphKind[reflect.TypeOf(&kapi.Service{})] = ServiceGraphKind
+	kindToGraphKind[reflect.TypeOf(&kapi.Pod{})] = PodGraphKind
+	kindToGraphKind[reflect.TypeOf(&kapi.ReplicationController{})] = ReplicationControllerGraphKind
+}
+
+func GetUniqueNamespaceNodeName(obj runtime.Object) UniqueName {
```
Comment (ncdc) [inline pkg/api/graph/types.go]: I'd probably call this GetUniqueNamespaceScopedNodeName to align with GetUniqueRootScopedNodeName

File: pkg/api/graph/types.go [hunk 4]
```diff
@@ -47,6 +51,72 @@ const (
...
+	objType := reflect.TypeOf(obj)
+	graphKind, exists := kindToGraphKind[objType]
+
+	if !exists {
+		panic(fmt.Sprintf("no graphKind registered for %v", objType))
+	}
+
+	meta, err := kapi.ObjectMetaFor(obj)
+	if err != nil {
+		panic(err)
```
Comment (ncdc) [inline pkg/api/graph/types.go]: Is there any time that an error here is valid and recoverable?

File: pkg/api/graph/types.go [hunk 5]
```diff
@@ -47,6 +51,72 @@ const (
...
+	Node
+	*image.ImageStream
+}
+
+func (n ImageStreamNode) Object() interface{} {
+	return n.ImageStream
+}
+
+func (n ImageStreamNode) String() string {
+	return fmt.Sprintf("<ImageStream %s/%s>", n.Namespace, n.Name)
```
Comment (ncdc) [inline pkg/api/graph/types.go]: Do you want to be consistent with the other funcs as far as capitalization goes? ImageStream -> imagestream? Personally I think CamelCase looks better (easier to distinguish separate terms).

File: pkg/api/graph/types.go [hunk 6]
```diff
@@ -124,13 +206,17 @@ func (n ImageStreamTagNode) Object() interface{} {
...
-	return fmt.Sprintf("<ImageStream %s/%s:%s>", n.Namespace, n.Name, n.Tag)
+	return fmt.Sprintf("<imagestream %s/%s:%s>", n.Namespace, n.Name, n.Tag)
 }
 
 func (*ImageStreamTagNode) Kind() int {
 	return ImageStreamTagGraphKind
 }
 
+func DockerImageRepositoryNodeName(o image.DockerImageReference) UniqueName {
+	return UniqueName(fmt.Sprintf("%d|%s/%s/%s:%s", DockerRepositoryGraphKind, o.Registry, o.Namespace, o.Name, o.Tag))
```
Comment (ncdc) [inline pkg/api/graph/types.go]: You'll probably want to call `o.String()` (see https://github.com/openshift/origin/blob/master/pkg/image/api/helper.go#L128), as `o.Tag` can be empty while `o.ID` is populated.

File: pkg/api/graph/types.go (hunk 2 above)
Comment (deads2k) [inline pkg/api/graph/types.go]: > Honestly, that's a pretty weak argument.  Test cases can do that too.

I like the way the code falls out better.  It makes a for a smooth transition to a metadata object if required.  It makes for an easy refactor into separate packages registering with one package to rule them all (a la scheme/codec/conversions).  It allows for the simple scheme coverage check without having to create partial/invalid objects. 

In any case, neither here nor there: fell out during another refactor here: https://github.com/openshift/origin/pull/3382.

**Comment by ncdc** [comment]
@deads2k are you planning on closing this in favor of #3382? Should I comment there instead?

**Comment by deads2k** [comment]
> @deads2k are you planning on closing this in favor of #3382? Should I comment there instead?

I had figured this was easier and would let me move on with adding secrets and serviceaccounts, but it seems to not be the case.  I'll focus on #3382 instead.
